### PR TITLE
Parallelize block_logical_coords with OpenMP if enabled

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

The majority of time spent in interpolation and loading numeric initial data is spent in `block_logical_coordinates`, mapping
all grid points through the domain. It's very easily parallelized though (literally 1 line of OpenMP for perfect strong scaling).

OpenMP is disabled in our charm++ execs unless explicitly requested, so this doesn't have an effect there. It currently only applies to pybindings and the exporter lib. However, we are loading numeric initial data on a nodegroup, so we may even want to enable the OpenMP parallelization there. I haven't tested it yet though.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: cmake/SetupPch.cmake
Comment (wthrowe) [inline cmake/SetupPch.cmake]: Have you measured the overhead of this?  I would not expect running OpenMP sections with a single thread to be as fast as compiling them without OpenMP, but I don't know if the change is significant for us.

**Review by wthrowe — CHANGES_REQUESTED** [review]
The code looks fine, aside from the things CI says.  I am worried about performance implications if you haven't benchmarked, though.

File: cmake/SetupPch.cmake
Comment (nilsvu) [reply cmake/SetupPch.cmake]: I added a benchmark with some results. I didn't find any measurable overhead when running on a single thread with OpenMP linked.

Comment (wthrowe) [reply cmake/SetupPch.cmake]: Sounds good, then.

**Review by wthrowe — APPROVED** [review]

**Review by wthrowe — CHANGES_REQUESTED** [review]
Based on CI, this appears to have caused a significant slowdown in the Unit.NumericalAlgorithms.Interpolator.InterpolateEventNoInterpolator test.

**Comment by knelli2** [comment]
Let's hold off on merging this until @nilsvu does some testing on mbot. It's possible the slow interpolation is a CaltechHPC thing and not our algorithm

**Comment by nilsvu** [comment]
This fixup should fix it @wthrowe.

@knelli2 This is independent and can be merged. This PR adds the OMP parallelization for pybindings and exporter. Probably won't be used in our parallel executables.

File: cmake/SpectreAddCatchTests.cmake
Comment (wthrowe) [inline cmake/SpectreAddCatchTests.cmake]: I don't like this solution, but I haven't thought of a better one so far.  This means that running the test executables directly isn't going to do what they are intended to do or what people expect, and there's not going to be an easy way for people to figure out what's going on.

I've been confused in similar ways in the past by tests that do benchmarking, but at least for those the benchmarking code is obvious in the test.

Comment (nilsvu) [reply cmake/SpectreAddCatchTests.cmake]: The alternative is to disable multithreading in the test main function, which is very easy to do (I can push a quick commit tomorrow). I didn't do that because I wanted to control the benchmark with the environment variable, but it's certainly more predictable. I can add a line to the benchmark to set the num threads at compile time.

Comment (nilsvu) [reply cmake/SpectreAddCatchTests.cmake]: pushed

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
I'm blocking so I can look at this. A few quick notes:

1. I remember we discussed in the past not linking the entire code base against OpenMP.
2. It was definitely not communicated at all that everything would depend on OpenMP. I'm not sure I like the idea of require OpenMP everywhere. 
3. This also seems much more complicated than originally advertised, but I'll need to look more carefullly..

File: src/Domain/BlockLogicalCoordinates.cpp
Comment (nilsdeppe) [inline src/Domain/BlockLogicalCoordinates.cpp:142]: Did you ever try just doing:
```cpp
#include <thread>

const auto lambda = [&x, time, &functions_of_time](const size_t start, const size_t end) {
  for (size_t s = start; s < end; ++s) {
    tnsr::I<double, Dim, Frame> x_frame(0.0);
    for (size_t d = 0; d < Dim; ++d) {
      x_frame.get(d) = x.get(d)[s];
    }
    // Check which block this point is in. Each point will be in one
    // and only one block, unless it is on a shared boundary.  In that
    // case, choose the first matching block (and this block will have
    // the smallest block_id).
    for (const auto& block : domain.blocks()) {
      std::optional<tnsr::I<double, Dim, ::Frame::BlockLogical>> x_logical =
          block_logical_coordinates_single_point(x_frame, block, time,
                                                 functions_of_time);
      if (x_logical.has_value()) {
        // Point is in this block.  Don't bother checking subsequent
        // blocks.
        block_coord_holders[s] = make_id_pair(domain::BlockId(block.id()),
                                              std::move(x_logical.value()));
        break;
      }
    }
  }
};

if (num_threads.value_or(0) == 0) {
  lambda(0, num_pts);
} else {
  std::vector<std::thread> threads(num_threads.value());
  // Spawn threads to do the interpolation.
  for (auto& thread : threads) {
    // Note: you'll want to do the math so that the remainder grid points are
    // approximately evenly distributed.
    thread = std::thread{lambda, this_thread_start, this_thread_end};
  }
  // Wait for threads to finish the interpolation.
  for (auto& thread : threads) {
    thread.join();
  }
}
```
I'm not sure if you'll get "perfect strong scaling", and it's more than "just a pragma", but it also doesn't require _any_ changes to other files, no env vars, etc. Plus, this doesn't even require a compiler that supports OpenMP (for a long time Clang didn't!)

Comment (nilsvu) [reply src/Domain/BlockLogicalCoordinates.cpp:142]: Thanks for offering the suggestion! Seems very reasonable. OpenMP is so easy though :D The parallelization of `block_logical_coordinates` is indeed just one line ("as advertised"). The parallelization of the exporter lib is a bit more complicated (still very easy though), but that's a separate lib that's only used for pybindings and exporter lib.

I limited linking OpenMP out of an abundance of caution before. I think that's still possible by compiling a parallelized version of `block_logical_coordinates` into the exporter lib, and leaving the domain lib untouched. I can try that.

Note that this parallelization speeds up the exporter by a factor 10 at least, and this is actually very noticeable when interpolating data for plots or analysis, and essential for importing data into another code. That's why I'd like to find a path for merging this.

I think independent of this we should disable OpenMP multithreading in our executables, so no dependencies accidentally try to do multithreading (like OpenBLAS, which we already disable, or libsharp). I think I'll just add a line `OMP_NUM_THREADS=1` to our submit script template, just to be safe.

Comment (nilsdeppe) [reply src/Domain/BlockLogicalCoordinates.cpp:142]: You're welcome! I don't see how you get to 1 line. The OpenMP-specific changes in BlockLogicalCoordinates are
```cpp
#ifdef _OPENMP
#include <omp.h>
#endif  // _OPENMP

// Ignore OpenMP pragmas when OpenMP is not enabled
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wunknown-pragmas"

  // Resolve number of threads to use in OpenMP parallelization
#ifdef _OPENMP
  const size_t resolved_num_threads =
      num_threads.value_or(omp_get_max_threads());
#else
  if (num_threads.has_value()) {
    ERROR("OpenMP is not available, so num_threads cannot be specified.");
  }
#endif  // _OPENMP

#pragma omp parallel for num_threads(resolved_num_threads)

#pragma GCC diagnostic pop
```
I count ~17 lines of code. While not huge, these are somewhat annoying "if" macros, which are hard to read. Macros also do not play well with C++ modules, so this adds a potential future challenge. I've also not had great success with the `-Wunknown-pragmas` ignoring on mbot (I haven't looked into why) where the warnings are printed. Furthermore, as Will pointed out, now the tests have to have an env variable set.

Additionally, putting another parallelization method into the code that literally over 99% of the source tree doesn't use is something I really don't like. Interoperability with other frameworks is always difficult, and once we merge something like this, it becomes someone else's problem if they find issues. I've had this happen several times and it's cost me a lot of time. During our telecon the discussion that led to "yes, go ahead with OpenMP" was restricted to the exporter and clearly stated that nothing else should depend on OpenMP. Thus, I currently don't see a path forward with linking OpenMP into the entire code base that I would be comfortable with. C++ threads are a language feature and control of them is very easy. I'll note that the diff I gave you has roughly the same number of lines changed in the file, the lines are contained to 1 place (excluding the include), there are no macros (good for C++ modules compatibility), and still runs in parallel. You might still get a 10x speedup, and the entire diff could shrink by a non-trivial amount. Certainly the CMake would be a lot easier.

Adding the line to submit scripts seems fine. We currently build libsharp without OpenMP and at least I try to use serial OpenBLAS implementations in addition to us disabling it explicitly for OpenBLAS.

Comment (nilsvu) [reply src/Domain/BlockLogicalCoordinates.cpp:142]: As noted above I think it's possible to remove the OpenMP linking from the domain lib (and thus also from the PCH and the rest of the code). I may not have written this clearly above. I can move the loop with its parallelization into the exporter lib and only call into the domain lib in the body of the loop. Then there's no parallelization in the domain lib at all.

Comment (nilsdeppe) [reply src/Domain/BlockLogicalCoordinates.cpp:142]: I think if you restrict the OMP pragmas to only the exporter lib it will be fine. Given the way Kokkos works, I will likely have Kokkos run in OpenMP mode while I debug our implementation before trying to run it on a GPU. That means if _any_ code that we use in our executables uses OpenMP, there's a problem. That means the domain files cannot have any OpenMP includes or directives, etc. in them.

**Comment by nilsvu** [comment]
Updated. Please take another look @wthrowe @nilsdeppe.

**Review by nilsdeppe — APPROVED** [review]
Yep, this looks good to me! Thank you!

**Review by wthrowe — APPROVED** [review]

**Pull request merged**
